### DISCUSSION TITLE
How to access kernels’ parameter values in multioutput Gaussian processes and multitask Gaussian processes

### DISCUSSION THREAD

**Discussion by GHU2021** [cornellius-gp/gpytorch#1624] [Q&A] [answered]
Hi,

If we use ExactGPModel with for example ScaleKernel and RBFKernel, we can use the following commands to access kernels’ parameter values (values of lengthscale, outputscale) and noise variance:
model.covar_module.base_kernel.lengthscale.item()
model.covar_module.raw_outputscale
model.likelihood.noise.item()

Similarly, if we use MultitaskGPModel or BatchIndependentMultitaskGPModel, how can we access to kernels' parameter values (e.g., if we define the addition of RBF kernel and linear kernel) in these models after we optimized these models?

I appreciate you help.

**Comment by gpleiss** [comment, marked as answer]
Drop the `.item()`. E.g.

```python
model.covar_module.base_kernel.lengthscale
model.covar_module.raw_outputscale
model.likelihood.noise
```

**Reply by GHU2021** [reply]
Dear Dr. Pleiss,
Thank you for your response.
.item() is necessary because I need to have values of kernel parameters and write them in a file.
Assume that I use MultitaskGPModel as follows:

class MultitaskGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood, L, M):
        super(MultitaskGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.MultitaskMean(gpytorch.means.ConstantMean(), num_tasks=L )
        self.covar_module = gpytorch.kernels.MultitaskKernel(
            gpytorch.kernels.RBFKernel(active_dims=list(range(0,M)))+gpytorch.kernels.RBFKernel(active_dims=list(range(M,2*M))), num_tasks=L, rank=1
        )
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultitaskMultivariateNormal(mean_x, covar_x)

Then, to check lengthscale value of the first RBF kernel, I should use 
model.covar_module.data_covar_module.kernels[0].lengthscale.item()
and to access to noise variance of Multitask GP, I should use
model.likelihood.noise.item()
